Helper/Webhook.php

```diff
@@ -0,0 +1,1883 @@
+                sprintf("Notification %s is processed", $notification->getEntityId())
+            );
+
+            return true;
+        } catch (\Exception $e) {
+            $this->updateNotification($notification, false, false);
+            $this->handleNotificationError($notification, $e->getMessage());
+            $this->logger->addAdyenNotificationCronjob(
+                sprintf(
+                    "Notification %s had an error: %s \n %s",
+                    $notification->getEntityId(),
+                    $e->getMessage(),
+                    $e->getTraceAsString()
+                )
+            );
+
+            return false;
+        }
+    }
+
+    /**
+     * Remove OFFER_CLOSED and AUTHORISATION success=false notifications for some time from the processing list
+     * to ensure they won't close any order which has an AUTHORISED notification arrived a bit later than the
+     * OFFER_CLOSED or the AUTHORISATION success=false one.
+     * @param Notification $notification
+     * @return bool
+     */
+    public function shouldSkipProcessingNotification(Notification $notification): bool
+    {
+        switch ($notification->getEventCode()) {
+            case Notification::OFFER_CLOSED:
+                // OFFER_CLOSED notifications needs to be at least 10 minutes old to be processed
+                $offerClosedMinDate = new \DateTime('-10 minutes');
+                $createdAt = \DateTime::createFromFormat('Y-m-d H:i:s', $notification->getCreatedAt());
+                $minutesUntilProcessing = ($createdAt->format('U') - $offerClosedMinDate->format('U')) / 60;
+
+                if ($minutesUntilProcessing > 0) {
+                    $this->logger->addAdyenNotificationCronjob(
+                        sprintf(
+                            'OFFER_CLOSED notification (entity_id: %s) for merchant_reference: %s is skipped! Wait %s minute(s) before processing.',
+                            $notification->getEntityId(),
+                            $notification->getMerchantReference(),
+                            $minutesUntilProcessing
+                        )
+                    );
+
+                    return true;
+                }
+
+                break;
+            case Notification::AUTHORISATION:
+                // Only delay success=false notifications processing
+                if ($notification->isSuccessful()) {
+                    // do not skip this notification but process it now
+                    return false;
+                }
+
+                // AUTHORISATION success=false notifications needs to be at least 10 minutes old to be processed
+                $authorisationSuccessFalseMinDate = new \DateTime('-10 minutes');
+                $createdAt = \DateTime::createFromFormat('Y-m-d H:i:s', $notification->getCreatedAt());
+
+                $minutesUntilProcessing = ($createdAt->format('U') - $authorisationSuccessFalseMinDate->format('U')) / 60;
```

Comment:
Can we consolidate the minutesUntilProcessing calc so that we don't repeat the code for both OFFER_CLOSED and AUTHORISATION?

Reply:
f3950c1